File: 方子樽/2022.01.28/ValidMountainArray.js
/**
 * @param {number[]} arr
 * @return {boolean}
 */
const validMountainArray = function (arr) {
    let isIncreasing, isDecreasing = false;
    let changeTimes = 0

    if (arr.length < 3) return false;

    for (let i = 0; i < arr.length; i++) {
        if (changeTimes > 1) return false;

        if (arr[i] < arr[i + 1]) {
            if (isDecreasing) changeTimes++;
            isIncreasing = true;
            isDecreasing = false;
        }
        else if (arr[i] > arr[i + 1]) {
            if (isIncreasing) changeTimes++;
            isDecreasing = true;
            isIncreasing = false;
        }
        else if (arr[i] === arr[i + 1]) {
            return false
        }
    }
    //last one must be isDecreasing
    return (changeTimes === 0) ? false : isDecreasing
}

File: 許靜玟/2022.01.28/ValidMountainArray.cpp
class Solution {
public:
    bool validMountainArray(vector<int>& arr) {
        if(arr.size() == 1) return 0;
        int maxPosition = max_element(arr.begin(),arr.end()) - arr.begin();
        for(int i=0; i< arr.size()-1; i++)
        {
            if(i < maxPosition)
            {
                if(arr[i] >= arr[i+1] || maxPosition == arr.size()-1) return false;
            }
            else
            {
                if(arr[i] <= arr[i+1] || maxPosition == 0) return false;
            }
        }
        return true;
    }
};